Load screen + Terrain load performance increase

Triangle345 opened the pull request:
- Create a height map upon very first run. This map gets loaded into memory and parsed which tremendously increases the speed of loading the terrain map.

- Created loading screen to easily monitor longer than normal load times

On very first vanilla load without the binary map, a map will be made. This will take the normal long time. However; every time after that; it will display the loading screen and load much faster. As long as the binary map file is present, we don't need to ask the Terrain to calculate the height.

Will continue to clean this up. Just making an initial check in.

Commit:
- Upon first ever load (for debug purposes now) we create a binary height map which gets loaded into memory and parsed. This tremendously increases the speed of parsing the terrain map.
- Created loading screen to easily monitor longer than normal load times

Triangle345 commented:
I may have discovered a few issues with my compression algorithm. Let me commit a fix before we merge this in.

pvutov commented on src/FieldWarning/Assets/Model/Game/Loading.cs:
```diff
@@ -0,0 +1,51 @@
+﻿using System;
```
Missing license

Triangle345 replied:
Thanks for the review!

pvutov replied:
The new files are still missing the license header (you can copy it from any existing code file in the project).

Triangle345 replied:
apologies... not sure what happened there, thought I put the header up everywhere.

pvutov commented on src/FieldWarning/Assets/LoadingScreen.cs:
```diff
@@ -0,0 +1,61 @@
+    private Slider slider;
+    private Text descLbl;
+
+    static public Queue<Loading> workers = new Queue<Loading>();
```
code style: _underscoreNameForPrivateVariable, PascalCaseForPublic (maybe S prefix for static)

pvutov commented on src/FieldWarning/Assets/LoadingScreen.cs:
```diff
@@ -0,0 +1,61 @@
+    }
+
+
+    void Update()
```
The unity inbuilt methods can be private (private void Start() etc)

pvutov commented on src/FieldWarning/Assets/LoadingScreen.cs:
```diff
@@ -0,0 +1,61 @@
+    void Update()
+    {
+        if (workers.Count > 0) 
+        {
```
We write the opening brace inline except for method and class definitions (kernel style). That said I think cuoshen and a lot of other people default to your style, so we could change that rule if you ask cuoshen and he prefers the style used here.

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -34,67 +36,226 @@ public class TerrainMap
     public const float BRIDGE_WIDTH = 3f * TerrainConstants.MAP_SCALE;
     public const float BRIDGE_HEIGHT = 1.0f; // temporary
 
+    public const string height_map_path = "Assets/Terrain/SampleHeight.txt";
```
HEIGHT_MAP_PATH (all caps constants)

pvutov commented on src/FieldWarning/Assets/Model/Game/Loading.cs:
```diff
@@ -0,0 +1,51 @@
+    }
+
+
+    public void Load()
```
As far as I can tell these methods are only used in TerrainMap, so they should be at least protected (if you go with inheritance).

Triangle345 replied:
actually this method Load() is used by LoadingScreen. It needs to be public. I can probably redesign it later to have lower visibility

pvutov commented on src/FieldWarning/Assets/Model/Game/Loading.cs:
```diff
@@ -0,0 +1,51 @@
+using System.Threading;
+using UnityEngine;
+
+public abstract class Loading
```
I would prefer to use composition instead of inheritance here. That's to say I'd prefer to have this as a normal class and have TerrainMap hold an instance of it as a variable. The calls below would then look like

`_loader.AddWorker(LoadHeightMap, "Loading height map");`

which I think is a lot clearer. In my experience inheritance hierarchies are just very hard to understand and change in the future. But you have worked more with this code so I will defer to your opinion if you strongly disagree. 

It would also help a lot to have comments for the class and methods about what their role is.

Triangle345 replied:
not a problem

pvutov commented on src/FieldWarning/Assets/Model/Game/Loading.cs:
```diff
@@ -0,0 +1,51 @@
+    public double percent_done = 0;
+    public bool finished = false;
+    public bool started = false;
+    public string name;
```
General trend here, try to make variables as private as possible, especially when they're not used by other classes anyway.

Triangle345 replied:
Thanks I will keep that in mind. These unfortunately all need to be public as they provide information about state and what the loader is doing.

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -34,67 +36,226 @@ public class TerrainMap
+
+                var bType = (byte)(BitConverter.ToSingle(type, 0) > waterHeight ? PLAIN : WATER);
+
+                var zEnd = z + len;
```
I feel like the names could be a lot more descriptive than `z` and `x` and `len`. Is it `fileLength`? Is `z` a coordinate? Is `zEnd` the highest allowed position on the z axis?

Also `var` is a bit overused, it's fine if the type is obvious but in the line `var zEnd = z + len;` this is not the case.

Commit:
-Fixed some bugs with the binary map compression
-Added a unit test for the map

Triangle345 commented:
Alright, I have updated and everything looks good. I incorporated @pvutov 's comments. I added a unit test to make sure my changes changes still contained the right terrain type.  Feel free to merge when ready.

pvutov commented on src/FieldWarning/Assets/Terrain/Scripts/TerrainMap.cs:
```diff
@@ -34,67 +36,314 @@ public class TerrainMap
+
+            if (_loader != null)
+                // this is our loading screen status
+                _loader.PercentDone = ((double)reader.BaseStream.Position / (double)reader.BaseStream.Length) * 100.0;
```
`Assets\Terrain\Scripts\TerrainMap.cs(270,25): error CS1061: 'Loading' does not contain a definition for 'PercentDone' and no accessible extension method 'PercentDone' accepting a first argument of type 'Loading' could be found (are you missing a using directive or an assembly reference?)`

I guess you already tried running this in unity and it worked, so perhaps some of the stuff pushed here is outdated?

Triangle345 replied:
correct, everything works locally on my end. Will need to checkout my repo and investigate

Commit:
- Re-checking in files

Triangle345 commented:
Not sure what happened there? Some of my files did not get committed? Well anyways, that would explain the lack of headers for the old files.. This should definitely work now. Let me know if it doesnt 👍
